plugins/gr-limesdr/lib/device_handler.cc

```diff
@@ -0,0 +1,753 @@
+}
+
+void device_handler::set_oversampling(int device_number, int oversample)
+{
+    switch (oversample) {
+    case 0:
+    case 1:
+    case 2:
+    case 4:
+    case 8:
+    case 16:
+    case 32: {
+        GR_LOG_DEBUG(d_debug_logger, "device_handler::set_oversampling(): ");
+        double host_value = device_handler::getInstance()
+                                .get_device(device_number)
+                                ->GetSampleRate(0, lime::TRXDir::Rx, 1);
+
+        if (device_handler::getInstance()
+                .get_device(device_number)
+                ->SetSampleRate(0, lime::TRXDir::Rx, 1, host_value, oversample) !=
+            lime::OpStatus::SUCCESS)
+            device_handler::getInstance().error(device_number);
+
+        GR_LOG_INFO(d_logger, fmt::format("Set oversampling: {:d}.", oversample));
+    } break;
+    default:
+        close_all_devices();
+        throw std::invalid_argument(
+            "device_handler::set_oversampling(): valid oversample values are: "
+            "0,1,2,4,8,16,32.");
+    }
+}
+
+double device_handler::set_rf_freq(int device_number,
+                                   lime::TRXDir direction,
+                                   int channel,
+                                   float rf_freq)
+{
+    double value = 0;
+
+    if (rf_freq <= 0) {
+        close_all_devices();
+        throw std::invalid_argument(
+            "device_handler::set_rf_freq(): rf_freq must be more than 0 Hz.");
+    } else {
+        GR_LOG_DEBUG(d_debug_logger, "device_handler::set_rf_freq(): ");
+        if (device_handler::getInstance()
+                .get_device(device_number)
+                ->SetFrequency(0, direction, channel, rf_freq) != lime::OpStatus::SUCCESS)
+            device_handler::getInstance().error(device_number);
+
+        value = device_handler::getInstance()
+                    .get_device(device_number)
+                    ->GetFrequency(0, direction, channel);
+
+        const std::array<std::string, 2> s_dir = { "RX"s, "TX"s };
+        GR_LOG_INFO(d_logger,
+                    fmt::format("RF frequency set [{:s}]: {:f} MHz.",
+                                s_dir[direction == lime::TRXDir::Tx ? 1 : 0],
+                                (value / 1e6)));
+    }
+
+    return value;
+}
+
+void device_handler::calibrate(int device_number,
+                               lime::TRXDir direction,
+                               int channel,
+                               double bandwidth)
+{
+    GR_LOG_DEBUG(d_debug_logger, "device_handler::calibrate(): ");
+
+    double rf_freq = device_handler::getInstance()
+                         .get_device(device_number)
+                         ->GetFrequency(0, direction, channel);
+    if (rf_freq > 31e6) // Normal calibration
+        device_handler::getInstance()
+            .get_device(device_number)
+            ->Calibrate(0, direction, channel, bandwidth);
+    else { // Workaround
```

Comment:
technically this is not a valid workaround and should not be performed, as the calibration results at different frequencies are incompatible.